expression/builtin_control.go

```diff
@@ -48,47 +53,312 @@ type caseWhenFunctionClass struct {
+	sc := b.ctx.GetSessionVars().StmtCtx
+	var (
+		condition, ret int64
+		isNull         bool
+		err            error
+	)
+	args, l := b.getArgs(), len(b.getArgs())
+	for i := 0; i < l-1; i += 2 {
+		condition, isNull, err = args[i].EvalInt(row, sc)
+		if err != nil {
+			return 0, isNull, errors.Trace(err)
+		}
+		if isNull || condition == 0 {
+			continue
+		}
+		ret, isNull, err = args[i+1].EvalInt(row, sc)
+		return ret, isNull, errors.Trace(err)
+	}
+	// when clause(condition, result) -> args[i], args[i+1]; (i >= 0 && i+1 < l-1)
+	// else clause -> args[l-1]
+	// If case clause has else clause, l%2 == 1.
+	if l%2 == 1 {
+		ret, isNull, err = args[l-1].EvalInt(row, sc)
+		return ret, isNull, errors.Trace(err)
 	}
+	return 0, true, nil
+}
+
+type builtinCaseWhenRealSig struct {
+	baseRealBuiltinFunc
+}
+
+// evalReal evals a builtinCaseWhenRealSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/case.html
+func (b *builtinCaseWhenRealSig) evalReal(row []types.Datum) (float64, bool, error) {
 	sc := b.ctx.GetSessionVars().StmtCtx
-	l := len(args)
+	var (
+		condition int64
+		ret       float64
+		isNull    bool
+		err       error
+	)
+	args, l := b.getArgs(), len(b.getArgs())
+	for i := 0; i < l-1; i += 2 {
+		condition, isNull, err = args[i].EvalInt(row, sc)
+		if err != nil {
+			return 0, isNull, errors.Trace(err)
+		}
+		if isNull || condition == 0 {
+			continue
+		}
+		ret, isNull, err = args[i+1].EvalReal(row, sc)
+		return ret, isNull, errors.Trace(err)
+	}
+	// when clause(condition, result) -> args[i], args[i+1]; (i >= 0 && i+1 < l-1)
+	// else clause -> args[l-1]
+	// If case clause has else clause, l%2 == 1.
+	if l%2 == 1 {
+		ret, isNull, err = args[l-1].EvalReal(row, sc)
+		if err != nil {
+			return ret, isNull, errors.Trace(err)
+		}
+		return ret, isNull, nil
+	}
+	return 0, true, nil
+}
+
+type builtinCaseWhenDecimalSig struct {
+	baseDecimalBuiltinFunc
+}
+
+// evalDecimal evals a builtinCaseWhenDecimalSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/case.html
+func (b *builtinCaseWhenDecimalSig) evalDecimal(row []types.Datum) (*types.MyDecimal, bool, error) {
+	sc := b.ctx.GetSessionVars().StmtCtx
+	var (
+		condition int64
+		ret       *types.MyDecimal
+		isNull    bool
+		err       error
+	)
+	args, l := b.getArgs(), len(b.getArgs())
 	for i := 0; i < l-1; i += 2 {
-		if args[i].IsNull() {
+		condition, isNull, err = args[i].EvalInt(row, sc)
+		if err != nil {
+			return nil, isNull, errors.Trace(err)
+		}
+		if isNull || condition == 0 {
 			continue
 		}
-		b, err1 := args[i].ToBool(sc)
-		if err1 != nil {
-			return d, errors.Trace(err1)
+		ret, isNull, err = args[i+1].EvalDecimal(row, sc)
+		return ret, isNull, errors.Trace(err)
+	}
+	// when clause(condition, result) -> args[i], args[i+1]; (i >= 0 && i+1 < l-1)
+	// else clause -> args[l-1]
+	// If case clause has else clause, l%2 == 1.
+	if l%2 == 1 {
+		ret, isNull, err = args[l-1].EvalDecimal(row, sc)
+		if err != nil {
```

Comment:
line 239 ~ line 242 is useless?

just use `return` in line 244 is enough?

so as the others.